apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+#include "Halide.h"
+
+#ifndef user_error
+#define user_error                Halide::Internal::ErrorReport(__FILE__, __LINE__, nullptr, Halide::Internal::ErrorReport::User)
+#endif
+
+#ifndef user_warning
+#define user_warning              Halide::Internal::ErrorReport(__FILE__, __LINE__, nullptr, Halide::Internal::ErrorReport::User | Halide::Internal::ErrorReport::Warning)
+#endif
+
+#ifndef user_assert
+#define user_assert(c)            _halide_internal_assertion(c, Halide::Internal::ErrorReport::User)
+#endif
+
+#ifndef internal_assert
+#define internal_assert(c)        _halide_internal_assertion(c, 0)
+#endif
+
+#ifndef internal_error
+#define internal_error            Halide::Internal::ErrorReport(__FILE__, __LINE__, nullptr, 0)
+#endif
+
+namespace Halide {
+namespace Internal {
+namespace Autoscheduler {
+
+std::map<std::string, Box> inference_bounds(const std::vector<Function> &functions,
+                                            const std::vector<Box> &output_bounds) {
+    std::vector<Func> funcs;
+    funcs.reserve(functions.size());
+    for (const auto &f : functions) {
+        funcs.push_back(Func(f));
+    }
+    return inference_bounds(funcs, output_bounds);
+}
+
+template <typename T>
+std::vector<int> sort_indices(const std::vector<T> &v) {
+    std::vector<int> idx(v.size());
+    std::iota(idx.begin(), idx.end(), 0);
+    std::sort(idx.begin(), idx.end(),
+         [&v](int i1, int i2) {return v[i1] < v[i2];});
+    return idx;
+}
+
+// Taken from src/AutoSchedule.cpp
```

Comment:
As per previous comment, we need to think about ways to avoid redundancy (or at least copy-n-paste) between different schedulers... I don't think it should block this PR, but perhaps it would be good to open a tracking issue of "stuff that needs to be reused by multiple scheduler plugins" so we know what the candidates are.

Reply:
I refactored these functions and expose them into Internal namespace.